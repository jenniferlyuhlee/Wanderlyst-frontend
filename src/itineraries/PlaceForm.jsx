import React, {useState} from "react";

/** PlaceForm component
 * Displays sub-form for adding associated places to itinerary
 * props (inherited from parent ItinForm): index, placeData, handlePlaceChange
 */
function PlaceForm({ index, placeData, handlePlaceChange }){
    return(
        <div className="place-form my-1 p-0">
            <h5 className = "mb-1 p-0">
                Place {index + 1}:
            </h5>
            <div className="form-floating mb-2 p-0">
                <input id = {`name${index+1}`}
                    className="form-control"
                    name="name"
                    placeholder="Place Name"
                    value={placeData.name}
                    onChange={(e) => handlePlaceChange(index, e)}
                    required
                />
                <label htmlFor={`name${index+1}`}>Place Name</label>
            </div>
            <div className="form-floating mb-2 p-0">
                <input id = {`address${index+1}`}
                    className="form-control"
                    name="address"
                    placeholder="Place Address"
                    value={placeData.address}
                    onChange={(e) => handlePlaceChange(index, e)}
                    required
                />
                <label htmlFor={`address${index+1}`}>Address</label>
            </div>
            <div className="form-group mb-2 p-0">
                <textarea
                    id ={`description${index+1}`}
                    className="form-control"
                    name="description"
                    rows = "3"
                    placeholder="Share what you did or can do here! Plus any time expectations, highlights, tips, important precautions etc. "
                    value={placeData.description}
                    onChange={(e) => handlePlaceChange(index, e)}
                    required
                />
            </div>
            <div className="form-floating mb-2 p-0">
                <input id = {`image${index+1}`}
                    className="form-control"
                    name="image"
                    placeholder="Image"
                    value={placeData.image}
                    onChange={(e) => handlePlaceChange(index, e)}
                />
                <label htmlFor={`image${index+1}`}>Image URL (optional)</label>
            </div>
        </div>
    );
}

export default PlaceForm;

