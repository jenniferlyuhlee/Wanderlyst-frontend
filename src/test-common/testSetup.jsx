import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };