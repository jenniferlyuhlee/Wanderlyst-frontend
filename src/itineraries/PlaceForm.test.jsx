import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter} from "react-router-dom";
import PlaceForm from "./PlaceForm";
import { expect, describe, test} from "vitest";

describe("TagsSelect sub-form component", () => {
    const mockPlaceData = {
        name: "",
        address: "",
        description: "",
        image: ""
    }
    const mockHandlePlaceChange = vi.fn();

    test("renders without crashing", () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <PlaceForm index={1} 
                                placeData={mockPlaceData}
                                handlePlaceChange={mockHandlePlaceChange}/>
                </UserProvider>
            </MemoryRouter>)
    });

    test("matches snapshot", () => {
        const {asFragment} = render(
            <MemoryRouter>
                <UserProvider>
                    <PlaceForm index={1} 
                                placeData={mockPlaceData}
                                handlePlaceChange={mockHandlePlaceChange}/>
                </UserProvider>
            </MemoryRouter>);
        expect(asFragment()).toMatchSnapshot();
    });

    test("updates form data and handles change", () => {
        const {getByLabelText, getByText} = render(
            <MemoryRouter>
                <UserProvider>
                    <PlaceForm index={1} 
                                placeData={mockPlaceData}
                                handlePlaceChange={mockHandlePlaceChange}/>
                </UserProvider>
            </MemoryRouter>);

        const name = getByLabelText("Place Name");
        const address = getByLabelText("Address");

        // fill in field and check that place change handler is called
        fireEvent.change(name, {target: {value: 'testplace name'}});
        expect(mockHandlePlaceChange).toHaveBeenCalled();

        fireEvent.change(address, {target: {value: 'test address'}});
        expect(mockHandlePlaceChange).toHaveBeenCalled();
    });
});