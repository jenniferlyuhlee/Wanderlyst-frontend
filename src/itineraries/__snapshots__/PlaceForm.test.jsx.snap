// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TagsSelect sub-form component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="place-form my-1 p-0"
  >
    <h5
      class="mb-1 p-0"
    >
      Place 2:
    </h5>
    <div
      class="form-floating mb-2 p-0"
    >
      <input
        class="form-control"
        id="name2"
        name="name"
        placeholder="Place Name"
        required=""
        value=""
      />
      <label
        for="name2"
      >
        Place Name
      </label>
    </div>
    <div
      class="form-floating mb-2 p-0"
    >
      <input
        class="form-control"
        id="address2"
        name="address"
        placeholder="Place Address"
        required=""
        value=""
      />
      <label
        for="address2"
      >
        Address
      </label>
    </div>
    <div
      class="form-group mb-2 p-0"
    >
      <textarea
        class="form-control"
        name="description"
        placeholder="Share what you did or can do here! Plus any time expectations, highlights, tips, important precautions etc. "
        required=""
        rows="3"
      />
    </div>
    <div
      class="form-floating mb-2 p-0"
    >
      <input
        class="form-control"
        id="image2"
        name="image"
        placeholder="Image"
        value=""
      />
      <label
        for="image2"
      >
        Image URL (optional)
      </label>
    </div>
  </div>
</DocumentFragment>
`;
